### PR TITLE
Add Cleanup Local to folders

### PR DESCRIPTION
**Pull request by JAAdventurer**
Adds "Cleanup Local" button to folder items on the dashboard, enabled if the local folder has any files/folders that the remote equivalent doesn't. Removes said files/folders from the local machine without a remote equivalent.

## Summary by CodeRabbit

* **New Features**
  * Added a **Cleanup Local** action for directories containing local-only files or folders.
  * Cleanup requires confirmation and is available only when eligible content is detected.
  * Added support for safely removing local-only content while preserving synchronized files.
  * Added server support for processing cleanup requests and reporting success or errors.
* **Bug Fixes**
  * Improved protection against invalid paths during local cleanup.
* **Tests**
  * Added coverage for cleanup eligibility, confirmation behavior, successful removal, and failure scenarios.

### CODE REVIEW

**Comment by JAAdventurer** [comment]
I'll address the issues in my fork when I get time tonight and send them down to the PR.

**Comment by JAAdventurer** [comment]
CodeRabbit comments addressed. @nitrobass24, have a look and see if this is something you want.

File: src/python/controller/command_pipeline.py [hunk 1]
```diff
@@ -310,6 +331,74 @@ def _handle_delete_remote(
...
+        ):
+            _notify_failure(
+                command,
+                f"Local contents of '{command.filename}' cannot be cleaned up in state {file.state!s}",
+            )
+            return False
+        if file.local_size is None:
+            _notify_failure(command, f"File '{command.filename}' does not exist locally")
+            return False
+        local_only_paths = _find_local_only_paths(file)
```
Comment (nitrobass24) [inline src/python/controller/command_pipeline.py:396]: **Blocking: a failed remote scan makes every child look local-only, and cleanup then deletes the whole folder's contents.**

`ModelUpdater` applies `latest_remote_scan.files` unconditionally (`model_updater.py:469-474`); the `failed` flag is only ever read to populate a status field. On a recoverable scanner error `ScannerProcess` publishes `ScannerResult(files=[], failed=True)`, so for that cycle the builder sees an empty remote tree and every child gets `remote_size = None`. The folder itself stays `DOWNLOADED` (local scan still supplies it), which passes the state gate above — so `_find_local_only_paths` returns every top-level child and `CleanupLocalProcess` guts the folder. Because the path list is re-derived here at dispatch time, this also hits a user who clicked during a healthy cycle if the command lands during the failed one.

Suggested guard before computing paths:
```python
if not pc.remote_scan_received or (pc.latest_remote_scan is not None and pc.latest_remote_scan.failed):
    _notify_failure(command, f"Cannot clean up '{command.filename}': remote listing unavailable")
    return False
if file.remote_size is None:
    _notify_failure(command, f"Cannot clean up '{command.filename}': folder no longer exists remotely")
    return False
```
The second check also covers the case where the remote folder was reaped (e.g. `auto_delete_remote`) — there, "cleanup" would be Delete Local minus the directory, which isn't what the button promises. Please mirror `remote_size != null` in `isCleanupLocalable` on the Angular side and add a unit test for a `remote_size=None` folder with local-only children being rejected.

(The root cause — `set_remote_files([])` on a failed scan — is a pre-existing bug worth its own issue, but this PR shouldn't depend on it being fixed.)

File: src/python/controller/delete/delete_process.py [hunk 2]
```diff
@@ -35,6 +39,34 @@ def run_once(self):
...
+            file_path = os.path.join(base_path, relative_path)
+            if not _is_contained(base_path, file_path):
+                self.logger.error(f"Path traversal blocked: {file_path} escapes {base_path}")
+                continue
+            if not os.path.exists(file_path):
+                self.logger.error(f"Failed to delete non-existing file: {file_path}")
+            elif os.path.isfile(file_path):
+                os.remove(file_path)
+            else:
+                shutil.rmtree(file_path, ignore_errors=True)
```
Comment (nitrobass24) [inline src/python/controller/delete/delete_process.py]: **Partial deletes are reported as success, and one `os.remove` failure aborts the rest of the batch.**

Two things in this loop:
- `os.remove(file_path)` (L65) is unwrapped. A `PermissionError`/`EBUSY`/`EROFS` on path *N* raises out of `run_once`, the child dies, and every remaining path in `relative_paths` is skipped. On the parent side that surfaces only as `logger.warning("Command process failed")` in `cleanup()` — no user notification, and `post_callback` has already fired.
- `shutil.rmtree(file_path, ignore_errors=True)` (L67) swallows every `OSError`: a read-only nested file leaves a half-deleted tree with zero log lines, and on 3.13 a **symlink to a directory** hits the "Cannot call rmtree on a symbolic link" path, which `ignore_errors` silences — so the symlink survives silently. `os.path.exists` (L62) is also `False` for a dangling symlink, so that's logged as "non-existing" and left in place.

Net effect: the user sees success, the rescan shows the junk still there, `isCleanupLocalable` stays true, and there's nothing in the log to explain it.

Suggested shape:
```python
failures: list[str] = []
for relative_path in self.__relative_paths:
    file_path = os.path.join(base_path, relative_path)
    try:
        if not _is_contained(base_path, file_path):
            raise OSError("escapes base path")
        if os.path.islink(file_path) or os.path.isfile(file_path):
            os.unlink(file_path)
        elif os.path.isdir(file_path):
            shutil.rmtree(file_path)  # let errors propagate
        elif not os.path.lexists(file_path):
            self.logger.warning(f"Skipping non-existing path: {file_path}")
    except OSError as e:
        self.logger.error(f"Failed to delete {file_path}: {e}")
        failures.append(relative_path)
if failures:
    raise OSError(f"Failed to clean up {len(failures)} path(s) under {base_path}: {failures}")
```
Checking `islink` *before* the containment check also fixes the case where a local-only symlink points outside the folder — today `realpath` resolution flags that as "traversal" and it's never removed. Please add real-tempdir tests (not mocked `realpath`) for: a `PermissionError` on path 1 still processing path 2, a symlink-to-dir, a dangling symlink, and `../outside.txt` staying untouched.

(`DeleteLocalProcess` above has the same `ignore_errors=True` blind spot; this PR copied it rather than fixing it — fine to leave, but worth a follow-up.)

File: src/angular/src/app/services/files/view-file.service.ts [hunk 3]
```diff
@@ -403,6 +420,8 @@ function createViewFile(modelFile: ModelFile, pairNameMap: Map<string, string>,
     modelFile.local_size,
     modelFile.remote_size,
   );
+  const isCleanupLocalable =
+    LOCAL_ACTION_STATUSES.includes(status) && modelFile.is_dir && hasLocalOnlyContent(modelFile);
```
Comment (nitrobass24) [inline src/angular/src/app/services/files/view-file.service.ts]: **`isCleanupLocalable` goes stale: `modelFilesEqual` never looks at `children`, but `children` is the only input to the new flag.**

`createViewFile` (L423) derives `isCleanupLocalable` from `hasLocalOnlyContent(modelFile)`, i.e. purely from the child tree. This comparator gates whether a ViewFile is rebuilt on an `updated` event, and it compares only the ten scalar fields. Any children-only change is treated as a no-op.

Two concrete failures:
- **Remote rename of a same-size file.** Folder `A` has `x.mkv` (100 B local+remote). Remote renames it to `y.mkv`. Model now has `x.mkv` (remote null) + `y.mkv` (local null). Parent `local_size`/`remote_size`/`state` are unchanged (dir size is the sum of children), so no update emits and the button stays **disabled** for exactly the orphan this feature exists to remove.
- **Zero-byte orphan (e.g. an empty `Sample/`).** After a cleanup, nothing on the parent changes, so the button stays **enabled** — and `file.component.ts:106-109` only clears `activeAction` on a `true→false` flip that never arrives, so the row's spinner wedges until navigation.

Cheapest sound fix is to compare the derived boolean rather than deep-walking the tree:
```ts
a.is_extractable === b.is_extractable &&
hasLocalOnlyContent(a) === hasLocalOnlyContent(b)
```
Please add a spec: emit a dir with a mirrored child, then re-emit the same dir with the child's `remote_size: null` and identical parent fields, and assert `isCleanupLocalable` flips to true. It fails today.

File: src/python/controller/command_pipeline.py [hunk 4]
```diff
@@ -310,6 +331,74 @@ def _handle_delete_remote(
...
+            if cleanup_path != _pc.local_path:
+                _pc.active_scan_process.force_scan()
+
+        command_wrapper = CommandProcessWrapper(process=process, post_callback=post_callback)
+        try:
+            command_wrapper.process.start()
+        except Exception as exc:
+            _notify_failure(command, f"Failed to start local cleanup for '{command.filename}': {exc!s}")
+            return False
+        self.active_command_processes.append(command_wrapper)
```
Comment (nitrobass24) [inline src/python/controller/command_pipeline.py:426]: **Non-blocking, but more important for cleanup than for its siblings:** `on_success` fires when the process *starts* (L136), so the only place a cleanup failure is ever observed is `cleanup()` (L458-462) — as a WARNING with just the process name, after `post_callback` has already fired. For DELETE_LOCAL/REMOTE that's tolerable because the file visibly disappears or doesn't. Cleanup is the first action where "nothing changed" is a plausible post-success state (N paths, partial failure — see my comment on `delete_process.py`), and the UI will show success while the rescan shows the junk still there.

Minimal ask for this PR: log at ERROR and include the file name (store it on `CommandProcessWrapper`), so the failure is at least findable. A user-facing channel (notification via the controller's `Notifier`, or a result queued back from the process) is a bigger change and fine as a follow-up issue — but worth opening, since without it the spinner in `file.component.ts:106-109` also has nothing to clear it on failure.

File: src/python/controller/command_pipeline.py [hunk 5]
```diff
@@ -310,6 +331,74 @@ def _handle_delete_remote(
...
+            _pc.local_scan_process.force_scan()
+            if cleanup_path != _pc.local_path:
+                _pc.active_scan_process.force_scan()
+
+        command_wrapper = CommandProcessWrapper(process=process, post_callback=post_callback)
+        try:
+            command_wrapper.process.start()
+        except Exception as exc:
+            _notify_failure(command, f"Failed to start local cleanup for '{command.filename}': {exc!s}")
+            return False
```
Comment (nitrobass24) [inline src/python/controller/command_pipeline.py:425]: **Nit: `except Exception` here is broader than it needs to be, and the process's pipes aren't closed on the failure path.**

The realistic failure from `start()` is `OSError` (fork/semaphore exhaustion — see #654 on the Alpine image). Catching `Exception` also turns programming errors (`TypeError` from a bad `set_mp_log_queue`, `AssertionError`) into a one-line toast, and on every retry the half-constructed `CleanupLocalProcess` and its `PipeStream` FDs are dropped without being closed. Suggest `except OSError as exc:` with `exc_info=True` on the log, and close the process's queues before returning.

File: src/python/controller/command_pipeline.py [hunk 6]
```diff
@@ -310,6 +331,74 @@ def _handle_delete_remote(
         command_wrapper.process.start()
         return True
 
+    def _handle_cleanup_local(
```
Comment (nitrobass24) [inline src/python/controller/command_pipeline.py:350]: **Test gap: this handler has no unit test.** `_find_local_only_paths` and `CleanupLocalProcess` are covered, and the integration test hits the happy path plus three failure strings, but the handler's own branches run only through that one path. Worth adding to `test_command_pipeline.py` (there's a template at `test_spawn_move_process_start_failure_leaves_no_stale_state`):

- **State gate**: a `DOWNLOADING` folder must be rejected with "cannot be cleaned up in state". This is the one that matters — lftp's `.lftp-pget-status`/partial files have `remote_size None` in the model, so if a state ever drops off the tuple, cleanup deletes an in-flight download.
- **Start failure**: patch `CleanupLocalProcess`, `start.side_effect = OSError`, assert `_notify_failure` gets "Failed to start local cleanup" and `active_command_processes == []`.
- **Staging branch**: with `use_staging` and the staging file present, assert the process gets `local_path == staging dir` and the post-callback forces both `local_scan_process` and `active_scan_process`; with it absent, only the local scan.
- **Cap deferral**: fill `active_command_processes` to the max, assert the command lands in `deferred` with no process created and no failure notified.

File: src/python/controller/delete/delete_process.py [hunk 7]
```diff
@@ -35,6 +39,34 @@ def run_once(self):
                 shutil.rmtree(file_path, ignore_errors=True)
 
 
+class CleanupLocalProcess(AppOneShotProcess):
+    """
+    Deletes a set of local paths, relative to a folder, that exist locally but not remotely.
+    Leaves the folder itself and any content that also exists remotely untouched.
+    """
```
Comment (nitrobass24) [inline src/python/controller/delete/delete_process.py:78]: **Docstring describes the caller's contract, not this class.** The process has no knowledge of remote state — it deletes whatever `relative_paths` it's handed. "Exist locally but not remotely" will rot the moment another caller reuses it. Suggest stating what *this* class guarantees:

> Deletes each of `relative_paths` beneath `<local_path>/<file_name>`. Each path is containment-checked against that base and skipped (with an error log) if it escapes. The folder itself is never removed. The caller (`CommandPipeline._handle_cleanup_local`) is responsible for choosing only local-only paths; `local_path` may be the staging dir.

Same note for `_handle_cleanup_local`'s one-line docstring in `command_pipeline.py` — this is the destructive path, and nothing there says the deletion rule is "any descendant with `remote_size is None`, stopping at the first local-only ancestor", or what happens in the staging case.

**Review by nitrobass24 — DISMISSED** [review]
Thanks for this — the feature is well-scoped, the pure helpers (`_find_local_only_paths` / `hasLocalOnlyContent`) mirror each other case-for-case with good tests, and the integration tests verify mirrored content survives via `dircmp`. I merged current `develop` into the branch (conflicts with this week's refactors) and pushed the result, so CI is green on the current head.

Requesting changes for three things that gate the merge, all in the inline comments:

1. **Data loss on a transient remote-scan failure** (`command_pipeline.py:371`) — a failed scan empties the remote tree for that cycle, so every child reads as local-only and cleanup guts the folder. Needs a guard on `remote_scan_received` / `latest_remote_scan.failed` / `file.remote_size is None`, mirrored in `isCleanupLocalable`.
2. **Partial deletes report success** (`delete_process.py:57-67`) — unwrapped `os.remove` aborts the batch; `rmtree(ignore_errors=True)` and the `realpath` check silently leave symlinks behind. Per-path error collection + raise a summary.
3. **Stale `isCleanupLocalable`** (`view-file.service.ts:423`) — `modelFilesEqual` doesn't consider `children`, so a same-size remote rename never enables the button and a zero-byte orphan wedges the row spinner.

The remaining comments (start-failure `except` scope, ERROR-level logging in `cleanup()`, handler unit tests, docstrings) are smaller and can ride along.

Once those land I'm happy to merge. Note for sequencing: this is a new feature, so it'll ship as v1.1.0 after the v1.0.4 cleanup release currently in #693.

**Comment by JAAdventurer** [comment]
I'll give it a look and address. I've got a couple other features I'm sitting on in my repo, so I'm not sure if you want to wait for those to bump the release or make other releases for those as well.

**Comment by nitrobass24** [comment]
Right now im just doing a bunch of refactors and clean up. Ill do minor version bumps after I get them tested on a live instance. I'm fine with whatever order you want to work them in. We can release them all at once if thats easier.

File: src/python/controller/command_pipeline.py (hunk 1 above)
Comment (JAAdventurer) [reply src/python/controller/command_pipeline.py:396]: Fixed in d65ab79: `_handle_cleanup_local` now rejects the command when `pc.remote_scan_received` is false or `pc.latest_remote_scan.failed` is true, and separately when `file.remote_size is None` (folder no longer exists remotely), extracted into a small `_remote_listing_failure_reason` helper to keep the handler under the complexity cap. Mirrored the second check in Angular's `isCleanupLocalable` (requires `modelFile.remote_size !== null`). Added unit tests for both guard branches, plus a live test against the dev SFTP setup: deleted the remote folder mid-cycle and confirmed cleanup is now rejected with "folder no longer exists remotely" instead of deleting the mirrored file.

File: src/python/controller/delete/delete_process.py (hunk 2 above)
Comment (JAAdventurer) [reply src/python/controller/delete/delete_process.py]: Fixed in d65ab79: `CleanupLocalProcess.run_once` now uses `os.unlink`/`shutil.rmtree` (no `ignore_errors`) with a per-path `try/except OSError`, collects failures instead of raising/continuing silently, and raises a single `RuntimeError` summarizing them after the loop so `propagate_exception()` surfaces it instead of the process reporting quiet success. Existing tests updated for the new unlink/rmtree calls and the raise-on-failure contract; added a new test asserting one failing path doesn't abort the rest of the batch. Verified live: a permission-locked local-only subfolder now fails loudly (ERROR log + surfaced failure) while an adjacent deletable local-only file in the same folder still gets removed.

File: src/angular/src/app/services/files/view-file.service.ts (hunk 3 above)
Comment (JAAdventurer) [reply src/angular/src/app/services/files/view-file.service.ts]: Fixed in d65ab79: `modelFilesEqual` now also compares `hasLocalOnlyContent(a) === hasLocalOnlyContent(b)` instead of only the ten scalar fields, so a children-only change (remote rename, orphaned empty dir) is no longer treated as a no-op. Added the spec you described: emits a dir with a mirrored child, re-emits with the child's `remote_size: null` and identical parent fields, asserts `isCleanupLocalable` flips to true.

File: src/python/controller/command_pipeline.py (hunk 4 above)
Comment (JAAdventurer) [reply src/python/controller/command_pipeline.py:426]: Fixed in d65ab79, generalized rather than special-cased to cleanup: added a `filename` field to `CommandProcessWrapper`, set at construction from `file.name` for all three command types (delete local/remote, cleanup), and `cleanup()` now logs at ERROR with that filename on any process failure (was WARNING with just the process class name). A user-facing failure channel is still a separate follow-up as you noted.

File: src/python/controller/command_pipeline.py (hunk 5 above)
Comment (JAAdventurer) [reply src/python/controller/command_pipeline.py:425]: Fixed in d65ab79: narrowed to `except OSError as exc:` with `exc_info=True` on the ERROR log, and added `command_wrapper.process.close_queues()` on the failure path (safe here since `close_queues()` only tears down the parent-side pipe/event objects created in `__init__`, not process join state).

File: src/python/controller/command_pipeline.py (hunk 6 above)
Comment (JAAdventurer) [reply src/python/controller/command_pipeline.py:350]: Fixed in d65ab79: added `TestHandleCleanupLocal` to `test_command_pipeline.py` covering the state gate (DOWNLOADING rejected), the new remote-scan guard (not-yet-received / failed scan / folder gone remotely), start failure (leaves no stale state, closes queues), the staging branch (both present and absent, asserting which scans get forced), and the concurrency-cap deferral.

File: src/python/controller/delete/delete_process.py (hunk 7 above)
Comment (JAAdventurer) [reply src/python/controller/delete/delete_process.py:78]: Fixed in d65ab79: reworded both docstrings to describe what each owns rather than the caller's contract. `CleanupLocalProcess`'s docstring now states the containment check, that the folder itself is never removed, and that the caller chooses the paths; `_handle_cleanup_local`'s docstring now states the deletion rule (first local-only ancestor, stopping recursion there) and the staging-dir behavior.

**Comment by JAAdventurer** [comment]
Pushed d65ab79 addressing all three blocking issues plus the four smaller ones. Replied inline on each thread with specifics; summary here:

- **Remote-scan guard** (`command_pipeline.py`): `_handle_cleanup_local` now rejects the command when the remote scan hasn't completed/failed, or when the folder no longer exists remotely, mirrored in Angular's `isCleanupLocalable`.
- **Partial deletes** (`delete_process.py`): `CleanupLocalProcess` now uses unwrapped `unlink`/`rmtree`, collects per-path failures, and raises a summary `RuntimeError` instead of reporting quiet success.
- **Stale `isCleanupLocalable`** (`view-file.service.ts`): `modelFilesEqual` now compares `hasLocalOnlyContent(a) === hasLocalOnlyContent(b)` so children-only changes trigger a rebuild.
- Narrowed the start-failure catch to `OSError` + close queues on failure.
- Generalized ERROR-level, filename-tagged logging to all `CommandProcessWrapper` failures (not just cleanup).
- Added unit tests for the handler's state gate, remote-scan guard, start failure, staging path, and concurrency cap.
- Reworded both docstrings you flagged.

Checks run locally (uv/pip weren't available here, so via Docker with the project's declared deps): `ruff check`/`ruff format --check` clean, `pyright` strict 0 errors, Python unit tests 998 passed (5 pre-existing multiprocessing/timing failures unrelated to this diff, confirmed present on a clean `develop` checkout too), integration tests 158 passed (extract-archive gaps are the documented local-only `rar`-tool gap; the cleanup-local integration tests themselves are gated on `SEEDSYNC_LIVE_SSH_TESTS` and run in CI), `ng lint` clean, `ng test` 577/577, `ng build --configuration production` succeeded.

Also field-tested live against the dev SFTP setup: cleanup removes local-only content and preserves mirrored content; deleting the remote folder mid-cycle now correctly blocks cleanup instead of gutting the mirrored file; a permission-locked local-only subfolder now fails loudly (ERROR log with filename) while an adjacent deletable local-only file in the same folder still gets removed.

**Comment by JAAdventurer** [comment]
@nitrobass24 Merged your PR in my repo, merging up your changes to my repo to address merge conflicts here, then will merge back down to this branch and this PR should be ready again.
